### PR TITLE
Integrate FIS SDK into Segmentation

### PR DESCRIPTION
**Pull request by ChaoqunCHEN**
Integrate FIS SDK into segmentation SDK to replace Instance ID SDK

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -117,6 +117,17 @@ jobs:
         - travis_retry ./scripts/pod_lib_lint.rb FirebaseInstanceID.podspec --platforms=ios --use-modular-headers
         - travis_retry ./scripts/pod_lib_lint.rb FirebaseInstanceID.podspec --platforms=tvos --use-modular-headers
         - travis_retry ./scripts/pod_lib_lint.rb FirebaseInstanceID.podspec --platforms=macos --use-modular-headers
+ 
+    - stage: test
+      if: type = cron
```
Comment (paulb777) [inline .travis.yml:122]: Delete this line so that travis runs on presubmits.

Also using travis is deprecated in favor of GitHub Actions.  See examples in .github/workflows

File: FirebaseInstallations/Source/Library/Public/FIRInstallations.h [hunk 2]
```diff
@@ -42,7 +42,7 @@ typedef void (^FIRInstallationsIDHandler)(NSString *__nullable identifier,
  * @param error The error when `tokenResult == nil` or `nil` otherwise.
  */
 typedef void (^FIRInstallationsTokenHandler)(
-    FIRInstallationsAuthTokenResult *__nullable tokenResult, NSError *__nullable error)
+FIRInstallationsAuthTokenResult *__nullable tokenResult, NSError *__nullable error)
```
Comment (paulb777) [inline FirebaseInstallations/Source/Library/Public/FIRInstallations.h:45]: Run `style.sh`. See https://github.com/firebase/firebase-ios-sdk#code-formatting

File: FirebaseSegmentation/Sources/FIRSegmentation.m [hunk 3]
```diff
@@ -0,0 +1,77 @@
+// Copyright 2019 Google
```
Comment (paulb777) [inline FirebaseSegmentation/Sources/FIRSegmentation.m:1]: Please make a separate PR for merging the existing already reviewed Segmentation code versus the new Installations code.

Comment (ChaoqunCHEN) [reply FirebaseSegmentation/Sources/FIRSegmentation.m]: Thanks for reviewing. Just done rebase. So now this pr only contains my changes.

File: FirebaseInstallations/Source/Library/Public/FIRInstallations.h (hunk 2 above)
Comment (ChaoqunCHEN) [reply FirebaseInstallations/Source/Library/Public/FIRInstallations.h:45]: done

File: FirebaseSegmentation/Tests/Unit/SEGContentManagerTests.m [hunk 4]
```diff
@@ -48,12 +57,22 @@ - (void)setUp {
   // Setup FIRApp.
   XCTAssertNoThrow([FIRApp configureWithOptions:[self FIRAppOptions]]);
   // TODO (mandard): Investigate replacing the partial mock with a class mock.
-  self.instanceIDMock = OCMPartialMock([FIRInstanceID instanceIDForTests]);
-  FIRInstanceIDResult *result = [[FIRInstanceIDResult alloc] init];
-  result.instanceID = @"test-instance-id";
-  result.token = @"test-instance-id-token";
-  OCMStub([self.instanceIDMock
-      instanceIDWithHandler:([OCMArg invokeBlockWithArgs:result, [NSNull null], nil])]);
+  //  self.instanceIDMock = OCMPartialMock([FIRInstanceID instanceIDForTests]);
```
Comment (paulb777) [inline FirebaseSegmentation/Tests/Unit/SEGContentManagerTests.m:60]: Delete

File: FirebaseSegmentation/Sources/SEGContentManager.m [hunk 5]
```diff
@@ -94,37 +94,51 @@ - (void)associateCustomInstallationIdentiferNamed:(NSString *)customInstallation
...
-    [strongSelf->_networkManager
-        makeAssociationRequestToBackendWithData:appAssociationData
-                                          token:strongSelf->_instanceIdentifierToken
-                                     completion:^(BOOL status,
-                                                  NSDictionary<NSString *, id> *result) {
-                                       // TODO...log, update database.
-
-                                       completionHandler(status, result);
-                                     }];
+    strongSelf->_instanceIdentifier = identifier;
```
Comment (maksymmalyhin) [inline FirebaseSegmentation/Sources/SEGContentManager.m:97]: Should we rename `instanceIdentifier` and `instanceIdentifierToken` as well?

File: FirebaseSegmentation/Tests/Unit/SEGContentManagerTests.m [hunk 6]
```diff
@@ -101,7 +122,7 @@ - (void)testAssociateCustomInstallationIdentifierSuccessfully {
 - (FIROptions *)FIRAppOptions {
   FIROptions *options = [[FIROptions alloc] initWithGoogleAppID:@"1:123:ios:123abc"
                                                     GCMSenderID:@"correct_gcm_sender_id"];
-  options.APIKey = @"correct_api_key";
+  options.APIKey = @"AIzaSyabcdefghijklmnopqrstuvwxyz1234567";
```
Comment (maksymmalyhin) [inline FirebaseSegmentation/Tests/Unit/SEGContentManagerTests.m:125]: nit: I would make the fake key looking more obviously like a fake one (more repeated laters/numbers, etc. ).

Comment (ChaoqunCHEN) [reply FirebaseSegmentation/Tests/Unit/SEGContentManagerTests.m:125]: I like it! thank you

File: FirebaseSegmentation/Sources/SEGContentManager.m [hunk 7]
```diff
@@ -63,29 +63,28 @@ - (instancetype)initWithDatabaseManager:databaseManager networkManager:networkMa
...
 }
 
 - (void)associateCustomInstallationIdentiferNamed:(NSString *)customInstallationID
                                       firebaseApp:(NSString *)firebaseApp
                                        completion:(SEGRequestCompletion)completionHandler {
-  // Get the latest instance identifier
-  if (![self instanceIDForApp:firebaseApp]) {
-    completionHandler(NO, @{kErrorDescription : @"InstanceID SDK not available"});
+  // Get the latest installation identifier
+  if (![self installationForApp:firebaseApp]) {
```
Comment (maksymmalyhin) [inline FirebaseSegmentation/Sources/SEGContentManager.m:74]: Will it make sense to declare a variable and call the method only once, like:
```
FIRInstallations *installation = [self installationForApp:firebaseApp];
if (installation == nil) {
...

[installation installationIDWithCompletion:...]
```

File: FirebaseSegmentation/Sources/SEGContentManager.m [hunk 8]
```diff
@@ -94,37 +93,51 @@ - (void)associateCustomInstallationIdentiferNamed:(NSString *)customInstallation
...
-                                          token:strongSelf->_instanceIdentifierToken
-                                     completion:^(BOOL status,
-                                                  NSDictionary<NSString *, id> *result) {
-                                       // TODO...log, update database.
-
-                                       completionHandler(status, result);
-                                     }];
+    strongSelf->_installationIdentifier = identifier;
+
+    [[self installationForApp:firebaseApp] authTokenWithCompletion:^(
```
Comment (maksymmalyhin) [inline FirebaseSegmentation/Sources/SEGContentManager.m:98]: use `strongSelf` here as well.

File: FirebaseSegmentation/Sources/SEGContentManager.m [hunk 9]
```diff
@@ -63,29 +63,28 @@ - (instancetype)initWithDatabaseManager:databaseManager networkManager:networkMa
...
-  if (![self instanceIDForApp:firebaseApp]) {
-    completionHandler(NO, @{kErrorDescription : @"InstanceID SDK not available"});
+  // Get the latest installation identifier
+  if (![self installationForApp:firebaseApp]) {
+    completionHandler(NO, @{kErrorDescription : @"Firebase Installations SDK not available"});
   }
   __weak SEGContentManager *weakSelf = self;
-  [[FIRInstanceID instanceID] instanceIDWithHandler:^(FIRInstanceIDResult *_Nullable result,
-                                                      NSError *_Nullable error) {
+  [[self installationForApp:firebaseApp] installationIDWithCompletion:^(NSString *__nullable identifier,
```
Comment (maksymmalyhin) [inline FirebaseSegmentation/Sources/SEGContentManager.m:78]: Using `self` here defeats the purpose of capturing `weakSelf`. `strongSelf` should be used in the block. To help catching such errors you may just move the code in the block to a separate method, so you have to call `strongSelf` only once and use regular `self` in the method. But it will be a bigger refactoring, so may be done separately.

File: FirebaseSegmentation/Sources/SEGContentManager.m [hunk 10]
```diff
@@ -94,37 +93,51 @@ - (void)associateCustomInstallationIdentiferNamed:(NSString *)customInstallation
...
+      if (!tokenResult || error) {
+        NSString *errorMessage = @"Error getting AuthToken.";
+        if (error) {
+          errorMessage = [errorMessage stringByAppendingString:error.description];
+        }
+        NSDictionary *errorDictionary = @{kErrorDescription : errorMessage};
+        completionHandler(NO, errorDictionary);
+        return;
+      }
+      strongSelf->_installationIdentifierToken = tokenResult.authToken;
```
Comment (maksymmalyhin) [inline FirebaseSegmentation/Sources/SEGContentManager.m:110]: If `strongSelf` is used here then `self` will be captured by this block. Lines 80-84 should be copied to the beginning of the block.

File: FirebaseSegmentation/Sources/SEGContentManager.m [hunk 11]
```diff
@@ -94,37 +93,51 @@ - (void)associateCustomInstallationIdentiferNamed:(NSString *)customInstallation
...
+
+      NSMutableDictionary<NSString *, NSString *> *appAssociationData =
+          [[NSMutableDictionary alloc] init];
+      [appAssociationData setObject:customInstallationID
```
Comment (maksymmalyhin) [inline FirebaseSegmentation/Sources/SEGContentManager.m:114]: nit: subscript syntax is preferred for dictionaries here and below, e.g. :
```
appAssociationData[forKey:kSEGCustomInstallationIdentifierKey] = customInstallationID;
```

I'm fine if it is addressed in the following PRs as it was introduced earlier.

File: FirebaseSegmentation/Tests/Unit/SEGContentManagerTests.m (hunk 4 above)
Comment (maksymmalyhin) [reply FirebaseSegmentation/Tests/Unit/SEGContentManagerTests.m:60]: Looks like it's still here ;)

File: FirebaseSegmentation/Tests/Unit/SEGContentManagerTests.m [hunk 12]
```diff
@@ -48,12 +57,22 @@ - (void)setUp {
...
+  //  result.instanceID = @"test-instance-id";
+  //  result.token = @"test-instance-id-token";
+  //  OCMStub([self.instanceIDMock
+  //      instanceIDWithHandler:([OCMArg invokeBlockWithArgs:result, [NSNull null], nil])]);
+
+  // Installations Mock
+  NSString *FID = @"fid-is-better-than-iid";
+  FIRInstallationsAuthTokenResult *FISToken =
+      [[FIRInstallationsAuthTokenResult alloc] initWithToken:@"fake-fis-token" expirationDate:nil];
+  self.installationsMock = OCMPartialMock([FIRInstallations installations]);
```
Comment (maksymmalyhin) [inline FirebaseSegmentation/Tests/Unit/SEGContentManagerTests.m:71]: I don't think a partial mock is needed here. I think we should replace it by a class mock:
```
self.installationsMock = OCMClassMock([FIRInstallations class]);
```

See an example of mocking FIRInstallations e.g. [here](https://cs.opensource.google/firebase-sdk/firebase-ios-sdk/+/master:Example/InstanceID/Tests/FIRInstanceIDTokenOperationsTest.m;l=409;drc=ef0f51e116cfe64d7cee1bc491cd09209a72218c). You'll need to stub `[FIRInstallations installationsWithApp:]` method instead `[FIRInstallations installation]` as you don't use a default instance.

As see []

File: FirebaseSegmentation/Tests/Unit/SEGInitializationTests.m [hunk 13]
```diff
@@ -43,6 +43,10 @@ - (void)tearDown {
   // class.
 }
 
+-(void)testCustomizedFirebaseAppIsUsedSuccess{
```
Comment (maksymmalyhin) [inline FirebaseSegmentation/Tests/Unit/SEGInitializationTests.m:46]: Implement?

**Review by maksymmalyhin — CHANGES_REQUESTED** [review]

File: FirebaseSegmentation/Sources/SEGContentManager.m (hunk 9 above)
Comment (ChaoqunCHEN) [reply FirebaseSegmentation/Sources/SEGContentManager.m:78]: Thank you maksym. I've noted down the issue. We'll do it in a separate PR.

File: FirebaseSegmentation/Sources/SEGContentManager.m (hunk 8 above)
Comment (ChaoqunCHEN) [reply FirebaseSegmentation/Sources/SEGContentManager.m:98]: will do in a separate PR.

File: FirebaseSegmentation/Sources/SEGContentManager.m (hunk 9 above)
Comment (ChaoqunCHEN) [reply FirebaseSegmentation/Sources/SEGContentManager.m]: done. Thank you!

File: FirebaseSegmentation/Sources/SEGContentManager.m [hunk 14]
```diff
@@ -63,69 +63,112 @@ - (instancetype)initWithDatabaseManager:databaseManager networkManager:networkMa
...
+  // Get the latest installation identifier
+  FIRInstallations *installation = [self installationForApp:firebaseApp];
+  if (installation == nil) {
+    completionHandler(NO, @{kErrorDescription : @"Firebase Installations SDK not available"});
   }
   __weak SEGContentManager *weakSelf = self;
-  [[FIRInstanceID instanceID] instanceIDWithHandler:^(FIRInstanceIDResult *_Nullable result,
-                                                      NSError *_Nullable error) {
+  [installation
+      installationIDWithCompletion:^(NSString *__nullable identifier, NSError *_Nullable error) {
```
Comment (maksymmalyhin) [inline FirebaseSegmentation/Sources/SEGContentManager.m:80]: nit: please use either `__nullable` or `_Nullable`. They are interchangeable but `_Nullable` seems to be wider used in Firebase codebase.

File: FirebaseSegmentation/Sources/SEGContentManager.m [hunk 15]
```diff
@@ -63,69 +63,112 @@ - (instancetype)initWithDatabaseManager:databaseManager networkManager:networkMa
...
+        SEGContentManager *strongSelf = weakSelf;
+        if (!strongSelf) {
+          completionHandler(NO, @{kErrorDescription : @"Internal Error getting installation ID."});
+          return;
+        }
+
+        [self associateInstallationWithLatestIdentifier:identifier
```
Comment (maksymmalyhin) [inline FirebaseSegmentation/Sources/SEGContentManager.m:87]: strongSelf

File: FirebaseSegmentation/Sources/SEGContentManager.m [hunk 16]
```diff
@@ -63,69 +63,112 @@ - (instancetype)initWithDatabaseManager:databaseManager networkManager:networkMa
...
+  __weak SEGContentManager *weakSelf = self;
+  [installation authTokenWithCompletion:^(FIRInstallationsAuthTokenResult *__nullable tokenResult,
+                                          NSError *_Nullable error) {
     SEGContentManager *strongSelf = weakSelf;
     if (!strongSelf) {
-      completionHandler(NO, @{kErrorDescription : @"Internal Error getting instance ID."});
+      completionHandler(NO, @{kErrorDescription : @"Internal Error getting installation token."});
       return;
     }
+    [self associateInstallationWithToken:tokenResult
```
Comment (maksymmalyhin) [inline FirebaseSegmentation/Sources/SEGContentManager.m:122]: strongSelf

File: FirebaseSegmentation/Sources/SEGContentManager.m [hunk 17]
```diff
@@ -63,69 +63,112 @@ - (instancetype)initWithDatabaseManager:databaseManager networkManager:networkMa
...
-                                       // TODO...log, update database.
-
-                                       completionHandler(status, result);
-                                     }];
-  }];
+    NSDictionary *errorDictionary = @{kErrorDescription : errorMessage};
+    completionHandler(NO, errorDictionary);
+    return;
+  }
+  self->_installationIdentifierToken = tokenResult.authToken;
```
Comment (maksymmalyhin) [inline FirebaseSegmentation/Sources/SEGContentManager.m:144]: Just `_installationIdentifierToken` can be used outside of a block. Adding explicit `self->` is required only in the blocks.

This is valid for other instance variables in the method.

File: FirebaseSegmentation/Tests/Unit/SEGContentManagerTests.m [hunk 18]
```diff
@@ -19,41 +19,63 @@
...
+  //  result.instanceID = @"test-instance-id";
+  //  result.token = @"test-instance-id-token";
+  //  OCMStub([self.instanceIDMock
+  //      instanceIDWithHandler:([OCMArg invokeBlockWithArgs:result, [NSNull null], nil])]);
+
+  // Installations Mock
+  NSString *FID = @"fid-is-better-than-iid";
+  FIRInstallationsAuthTokenResult *FISToken =
+      [[FIRInstallationsAuthTokenResult alloc] initWithToken:@"fake-fis-token" expirationDate:nil];
+  self.installationsMock = OCMPartialMock(
```
Comment (maksymmalyhin) [inline FirebaseSegmentation/Tests/Unit/SEGContentManagerTests.m:73]: As mentioned before, it doesn't look like a partial mock is needed here. Could you please replace `OCMPartialMock` with `OCMClassMock` (see the example provided earlier.

**Review by maksymmalyhin — CHANGES_REQUESTED** [review]
Thank you for the updates. Mostly LGTM after a couple corrections.

File: FirebaseSegmentation/Tests/Unit/SEGInitializationTests.m (hunk 13 above)
Comment (ChaoqunCHEN) [reply FirebaseSegmentation/Tests/Unit/SEGInitializationTests.m]: removed

File: FirebaseSegmentation/Sources/SEGContentManager.m (hunk 17 above)
Comment (ChaoqunCHEN) [reply FirebaseSegmentation/Sources/SEGContentManager.m:144]: Thank you!

File: FirebaseSegmentation/Tests/Unit/SEGContentManagerTests.m (hunk 18 above)
Comment (ChaoqunCHEN) [reply FirebaseSegmentation/Tests/Unit/SEGContentManagerTests.m]: Yes, done in this commit.

File: FirebaseSegmentation/Sources/SEGContentManager.m (hunk 11 above)
Comment (ChaoqunCHEN) [reply FirebaseSegmentation/Sources/SEGContentManager.m]: thank you

**Review by maksymmalyhin — APPROVED** [review]
Thank you for the fixes and the refactoring! LGTM
